<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Newton_aula.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Um exemplo de uso do método de Newton**

José Claudinei Ferreira

Universidade Federal de Alfenas

Considere a equação, ou o sistema de equações, $$\begin{cases}3x+5y&=&2\\\\\frac{x^2}{9}+\frac{y^2}{16}&=&4\end{cases}\tag{1}.$$

Lembrando um pouco de Geometria Analítica, temos que trata-se da intersecção de uma reta com uma elipse, o que pode produzir até 2 soluções $u=(x,y)$.

Vamos definir a função
$$f(u)=\begin{bmatrix}f_1(u)\\\\f_2(u)\end{bmatrix}=\begin{bmatrix}3x+5y-2\\\\\frac{x^2}{9}+\frac{y^2}{16}-4\end{bmatrix}$$ e a função $$g(u)=\frac{1}{2}\|f(u)\|^2=\frac{1}{2}(f_1(u)^2+f_2(u)^2)=\frac{1}{2}f(u)\cdot f(u),$$ em que $\cdot$ denota o produto escalar usual.

Então a equação $(1)$ pode ser vista na equivalência
$$f(u)=0	\Longleftrightarrow\min_u g(u)=0.$$

**Obs:** <font color=blue> Note que $$A=Jf(u)=\begin{bmatrix}3&5\\\frac{2x}{9}&\frac{y}{8}\end{bmatrix}$$  não tem inversa, quando $$det(A)=\frac{3}{8}y-\frac{10}{9}x=0.$$

<font color=blue> Isso pode fazer com que o método que queremos propor não funcione sempre; dependendo da condição inicial $u_0$.


Definições em Python:

In [29]:
import numpy as np     # Para usar álgebra linear.

def f(u):
  x=u[0]
  y=u[1]
  p=[3*x+5*y-2,x**2/9+y**2/16-4]
  return np.array(p)

In [30]:
f([1,1])               # Teste de função f(u), para u=(1,1).

array([ 6.        , -3.82638889])

In [31]:
def Jf(u):             # Matriz do jaconiano de f(u).
  x=u[0]
  y=u[1]
  p=[[3,5],[2*x/9,y/8]]
  return np.array(p)

In [32]:
Jf([1,1])

array([[3.        , 5.        ],
       [0.22222222, 0.125     ]])

In [33]:
def g(u):
  p=np.dot(f(u),f(u))/2  # np.dot(u,v) calcula o produto interno de u com v.
  return p

Vamos determinar uma curva $u(t)$ que minimiza $g(u)$, partindo de um ponto $u_0$,  e usando a equação diferencial $$\begin{cases}u'(t)&=&-\left[Jf(u(t))\right]^{-1}f(u(t))\\\\u(0)&=&u_0\end{cases}\tag{2},$$ ou
$$\begin{cases}Jf(u(t))\,u'(t)&=&-f(u(t))\\\\u(0)&=&u_0\end{cases}\tag{2*}.$$

Isso e o método de Euler dão origem ao método de Newton, que pode ser reescrito como $$u_{j+1}=u_j-h\left[Jf(u_j)\right]^{-1}f(u_j),$$ ou $$\begin{cases}Jf(u_j)v&=&-hf(u_j)\\\\u_{j+1}=u_j+v\end{cases},\tag{3}$$ para evitar o cáclulo da matriz inversa de $Jf(u_j)$.

Escolhendo, por exemplo $u_0=(1,1)$ vamos determinar $u_1$ com a expressão acima.

In [34]:
u0=np.array([1,1])
g(u0)

25.32062596450617

Tomando $h=1$ e o método de Euler $(3)$, para resolver numéricamente a Equação $(2)$, temos

In [35]:
h=1

v=np.linalg.solve(Jf(u0),-h*f(u0)) # resolve a equação Jf(u0)v=-hf(u0)
u1=u0+v
u1

array([ 28.00943396, -16.40566038])

In [36]:
g(u1)

4999.142735316382

o que nos diz que $h=1$ é um passo grande, porque $g(u_1)>g(u_0)$.

Escolhemos então $h=1/10$, e temos:

In [37]:
h=1/10

v=np.linalg.solve(Jf(u0),-h*f(u0)) # resolve a equação Jf(u0)v=-hf(u0)
u1=u0+v
u1

array([ 3.7009434 , -0.74056604])

In [38]:
g(u1)

17.566166537962253

o que nos diz que $h=1/10$ é um passo em que, porque $g(u_1)<g(u_0)$.

Vamos então calcular $u_2$

In [39]:
v=np.linalg.solve(Jf(u1),-h*f(u1)) # resolve a equação Jf(u0)v=-hf(u0)
u2=u1+v
u2

array([ 3.96790557, -1.00874334])

In [40]:
g(u2)

14.20136861225617

Vemos que $g(u_2)<g(u_1)<g(u_0)$ e parace que estamos no caminho certo.

Automatizando... para calcular $u_10$:

In [41]:
n=10
u=u2

for i in range(2,n):
  v=np.linalg.solve(Jf(u),-h*f(u)) # resolve a equação Jf(u0)v=-hf(u0)
  u=u+v

un=u
un

array([ 5.00051556, -2.18189521])

In [42]:
g(un)

2.6153672876468574

Vemos que $g(u_{10})<g(u_2)$, e parece que estamos próximos da solução de $f(u)=0$.

Aumentamos então $h$, agora $h=1$ e calculamos $u_{20}$.

In [43]:
n=10
h=1
u=un

for i in range(0,n):
  v=np.linalg.solve(Jf(u),-h*f(u)) # resolve a equação Jf(u0)v=-hf(u0)
  u=u+v

un=u
un

array([ 5.57810452, -2.94686271])

In [44]:
g(un)

0.0

Obtemos $u_{20}=(5.57810452, -2.94686271)$ com uma solução do problema dado.

Tento usar outra condição incial, ou outro $u_0$, para encontrar a outra raiz.

Para mais detalhes veja o arquivo deste [link](https://github.com/j-claudinei-f/j-claudinei-f/blob/main/Direcao_maior_decrescimento.ipynb).